### DISCUSSION TITLE
Session 12-march-night--looking-for-heart — Mar 12, 2026

### DISCUSSION THREAD

**Discussion by SolbiatiAlessandro** [SolbiatiAlessandro/cogames-autoresearch#9] [Show and tell]
## Session 12-march-night--looking-for-heart — Mar 12, 2026

**Direction:** Get agents to earn hearts (heart_amount > 0) and align junctions (aligned_by_agent > 0). Yesterday's session crashed with OOM before any experiments. Today's session repeats that plan with OOM fix in place.

**Prior findings:**
- Best known config (mar7): milestones_2 + role_conditional + penalize_vibe_change + miner + scout → score=304, but heart_amount=0, aligned_by_agent=0
- ROOT CAUSE: standalone `miner` variant sets heart_gained=-0.1 for ALL agents (punishes scouts/scramblers for hearts)
- SOLUTION: Remove standalone miner/scout. Use milestones_2:25 + role_conditional + penalize_vibe_change only.

**Plan:**
1. ✅ Exp 1 (5 min): milestones_2:25 + role_conditional + penalize_vibe, ent_coef=0.10 — DONE
2. Exp 2 (10 min): milestones_2:50 + role_conditional + penalize_vibe, ent_coef=0.10 — scale objective
3. Exp 3+ (20 min): Best config, longer training with higher LR

**Experiment Log:**

### Exp 1: milestones_2:25 + role_conditional + penalize_vibe_change, ent_coef=0.10, 5min
- **Result: BREAKTHROUGH** 🎉
- composite_score: 71.2
- heart_amount: **1.8** ← hearts crafted for the first time!
- aligned_by_agent: **0.1** ← aligners actually aligning junctions!
- cogs_junctions_held: 257.0 ← our team holds territory
- miner_gained: 0.3, aligner_gained: 0.4 — gear specialization working
- Diagnosis confirmed: removing standalone miner/scout fixed the heart production chain
- Commit: 5b28400

**INFRA:** OOM fix applied by Alessandro in commit 9dfaf64 (VECTOR_NUM_ENVS=64, VECTOR_NUM_WORKERS=8). CUDA A40 confirmed.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 01:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 03:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 05:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 07:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 09:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 11:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 13:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 15:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 17:05 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
## 🔍 Infra Deep-Dive: Why Sessions Keep Dying (Tashi analysis, Mar 12)

Alex asked me to look at this properly since it's the second night in a row the autoresearch session died mid-run. Here's what I found.

---

### The Pattern

| Session | Started | Died | Experiments | Death Mode |
|---------|---------|------|-------------|------------|
| mar-9-night | ~23:30 Mar 9 | ~03:00 Mar 10 | ~8 | "Machine crashed" (from discussion #5) |
| 11-march-night | ~22:00 Mar 10 | ~08:22 Mar 11 | 1 | OOM → fixed → died again (unknown) |
| 12-march-night | 23:16 Mar 11 | ~01:20 Mar 12 | 12 | Silent death — last commit 01:19, nothing after |

**Every session dies within 2-5 hours.** The 12-march session is especially painful — it had a genuine breakthrough (first hearts crafted, first junctions aligned) and then silently stopped.

---

### Root Cause Analysis

**The sessions run as: `claude --dangerously-skip-permissions` on a RunPod GPU pod.** There is no process supervisor. When Claude Code dies, nothing restarts it.

Likely kill reasons, ranked by probability:

#### 1. 🧠 Context window exhaustion (MOST LIKELY)
Claude Code has a ~200k token context window. Each experiment iteration involves:
- Reading program.md (~4k tokens)
- Reading knowledge/ files, prior discussions (~10-20k tokens)  
- Reading results.tsv + git log (~2-5k tokens)
- The agent's reasoning + code edits (~5-10k tokens)
- Running `uv run train.py` and reading output (~2-5k tokens)
- Git operations, logging (~1-2k tokens)

That's roughly **25-40k tokens per experiment**. With 200k context, you get **5-8 experiments** before Claude Code has to start compacting/summarizing context — and around **10-15 experiments** it hits the wall and either crashes or gets so compressed it can't function.

**This matches perfectly:** mar-9 got 8 experiments, 12-march got 12.

The `autoresearch_loop.sh` wrapper was designed to fix this — it spawns a NEW Claude Code instance per experiment (fresh context each time). But the sessions are being started with bare `claude --dangerously-skip-permissions` instead, so one long-running Claude Code process eats through context and dies.

#### 2. 💾 RunPod pod preemption
If using spot/interruptible instances, RunPod can kill your pod at any time. This would explain the "machine crashed" from mar-9-night. But it wouldn't explain the 11-march and 12-march deaths which pushed commits (so the pod was still alive, just Claude Code stopped).

#### 3. 🌐 DNS / API failures on RunPod
The `runpod_setup.sh` already fixes DNS (sets Google + Cloudflare resolvers, disables IPv6). But transient API failures could cause Claude Code to error out mid-conversation. Without a retry wrapper, one bad API call = dead session.

#### 4. 💰 API credit exhaustion
The autoresearch_loop.log from the mar-9 local run shows the loop dying with "Credit balance is too low" at experiment #1303-1304. Could be hitting Anthropic API limits.

---

### The Fix: Use autoresearch_loop.sh, Not Bare Claude

The `autoresearch_loop.sh` wrapper already solves most of these problems:
- ✅ **Fresh context per experiment** — spawns a new `claude --print` per iteration
- ✅ **OOM protection** — `check_memory()` between runs
- ✅ **Process cleanup** — kills stray training processes
- ✅ **Crash recovery** — 3-strike system with WhatsApp alerts
- ✅ **Checkpoint archiving** — saves models before resets

**But it has its own issues:**
1. The local version references `watchdog.sh` and `sync_notion.sh` which don't exist (see watchdog.log — endless "No such file or directory")
2. It needs to run ON the RunPod pod, not locally
3. The `--print` flag means no interactive Claude Code — just one-shot prompts. This is actually the right approach for overnight autonomy.

### Recommended Setup for Next Session

```bash
# On RunPod pod:
bash runpod_setup.sh                     # DNS fix, install tools
export ANTHROPIC_API_KEY=sk-...          # set API key
export PATH="$HOME/.local/bin:$PATH"     # ensure uv, gh, claude in PATH

# Create branch
cd ~/Projects/cogames-autoresearch
git checkout -b autoresearch/<branch>

# Run the loop wrapper (not bare claude!)
nohup bash autoresearch_loop.sh > autoresearch_loop.log 2>&1 &
discard=$!
echo "Loop PID: $discard"
```

### What Still Needs Fixing in autoresearch_loop.sh

1. **Remove references to `watchdog.sh` and `sync_notion.sh`** — they don't exist and spam the logs
2. **Add exponential backoff on API failures** — if Claude returns "Credit balance too low" or a network error, sleep longer between retries instead of burning through iterations in 1-second loops
3. **Add a heartbeat file** — write a timestamp to `.last_alive` every iteration so external monitoring (Tashi's cron) can distinguish "still running but slow" from "dead"
4. **Consider `screen` or `tmux`** — `nohup` is fragile; a detached tmux session survives SSH disconnects and lets you reattach to inspect

### Longer-Term: Make It Self-Healing

The dream setup:
- `autoresearch_loop.sh` runs inside `tmux` on the pod
- Each iteration spawns fresh Claude Code with `--print` (one-shot, no context accumulation)
- A simple systemd service or cron on the pod restarts the loop if it dies
- Tashi's heartbeat cron (running here on Mac) monitors commits and alerts Alex if no activity for >2h
- The loop writes to a `.heartbeat` file; if it's stale, Tashi could even SSH into the pod and restart it (if SSH is set up)

---

**TL;DR:** Sessions die because bare `claude` runs exhaust context in ~10-12 experiments. The loop wrapper (`autoresearch_loop.sh`) already fixes this by spawning fresh instances per experiment, but it's not being used on the pod. Fix the wrapper's broken script references, deploy it to the pod, and run it inside tmux.

**Comment by SolbiatiAlessandro** [comment]
## Infra Analysis: OOM Root Cause

**TL;DR: You do NOT need a bigger machine.** The A40 (45GB) is plenty.

### Root Cause

`cogames.train.train()` auto-detects CPU cores. On RunPod's 96-core machine:
- Auto-scales to **48 workers × 288 envs**
- Each worker loads Python + torch + cogames (~500MB each)
- 48 workers alone = ~24GB, plus 288 env observation buffers
- Result: 45GB/45GB OOM before a single training step

### Fix (commit 9dfaf64)

Hard-coded caps in `train.py`:
```python
VECTOR_NUM_ENVS = 64    # was auto-scaling to 288
VECTOR_NUM_WORKERS = 8  # was using all 48 physical cores
```

64 envs × 8 agents = 512 total agents. BPTT=128 → batch_size=65,536 → ~40MB obs buffer. Fits easily in A40.

### Why it broke again on March 11

The autoresearch agent modifies `train.py` freely between sessions. It likely overwrote or removed the `VECTOR_NUM_ENVS`/`VECTOR_NUM_WORKERS` caps. The README warning exists but the agent doesn't reliably read it.

### Recommended hardening

1. Move env/worker caps to `prepare.py` (agent only modifies `train.py`)
2. Or add a validation check that refuses to start if envs > 128 or workers > 16
3. The knowledge/memory_issues.md doc is good but needs to be enforced programmatically

— Tashi 🪷

**Comment by SolbiatiAlessandro** [comment]
## 🔧 Updated Diagnosis: System RAM OOM (not context window)

Alex confirmed RunPod shows **45GB/45GB system memory** — the pod is OOM-killed.

### Why This Happens

Each `uv run train.py` spawns a heavy subprocess tree:
- `pufferlib.PuffeRL` → vectorized environments → multiprocessing workers
- PyTorch training buffers (policy, optimizer state, rollout storage)
- `cogames` env instances (one per vectorized env)

When running **bare `claude`**, Claude Code runs experiments sequentially inside one long session. The problem: **child processes and GPU/CPU memory don't get fully released between experiments.** Specifically:

1. **Python multiprocessing workers** — `pufferlib` spawns worker processes for vectorized envs. `train.py` calls `subprocess.Popen` → `uv run python -c <train_script>`. When the TIME_BUDGET hits, it sends `SIGTERM` to the parent process, but **orphaned multiprocessing children can survive** (they're forked, not threads).

2. **PyTorch CUDA caching** — PyTorch's CUDA memory allocator aggressively caches GPU memory and doesn't return it to the OS. Across experiments this creeps up.

3. **No `gc.collect()` or `torch.cuda.empty_cache()`** — `train.py` doesn't clean up after itself. The subprocess exits, but Python's memory fragmentation means the OS may not reclaim everything.

4. **Claude Code itself** — Claude Code is a Node.js process that accumulates conversation context in RAM. After 12 experiments, its own memory footprint grows substantially.

The `autoresearch_loop.sh` wrapper **does** handle this — it has `cleanup_processes()` that explicitly `pkill`s stray `multiprocessing.spawn`, `train.py`, and `pufferlib` processes between experiments, plus memory checks. But it's not being used.

### The Fix (Two Options)

**Option A: Use `autoresearch_loop.sh` (recommended)**
Already has process cleanup + memory checks + fresh Claude per experiment. Needs these fixes first:
- Remove broken `watchdog.sh`/`sync_notion.sh` references
- Deploy to RunPod pod
- Run inside `tmux`

**Option B: Quick patch if you want to keep bare `claude`**
Add aggressive cleanup to `train.py` itself — at the end of `main()`:

```python
# Force cleanup after training
import gc
try:
    import torch
    if torch.cuda.is_available():
        torch.cuda.empty_cache()
        torch.cuda.reset_peak_memory_stats()
except: pass
gc.collect()

# Kill any orphaned multiprocessing workers
import signal, os
for proc_name in ["multiprocessing.spawn", "pufferlib"]:
    os.system(f"pkill -f {proc_name} 2>/dev/null || true")
```

And add to `program.md` instructions: after each experiment run, Claude should execute:
```bash
pkill -f "multiprocessing" 2>/dev/null; pkill -f "pufferlib" 2>/dev/null; sleep 5
```

**Option C (nuclear): Reduce env count**
The `[infra] fixing OOM` commit (9dfaf64) set `VECTOR_NUM_ENVS=64, VECTOR_NUM_WORKERS=8`. If memory is still tight, try `VECTOR_NUM_ENVS=32, VECTOR_NUM_WORKERS=4`. Halves the parallel env memory at the cost of ~2x slower data collection.

### Recommendation

Go with **Option A** — use the loop wrapper. It already solves this problem by design (fresh process per experiment + explicit cleanup). The bare `claude` approach will always eventually OOM because you can't fully control memory cleanup from inside a long-running Claude Code session.

For tonight's session:
1. SSH into the pod
2. `bash runpod_setup.sh`
3. Fix `autoresearch_loop.sh` (I can push the fix if you want)
4. `tmux new -s autoresearch`
5. `nohup bash autoresearch_loop.sh &` inside tmux
6. Detach and walk away

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 19:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 21:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-12 23:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-13 01:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-13 09:03 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-13 21:20 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-13 21:24 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
=== March 13 Update ===

**BEST RESULT:** commit 45956e8 - score **82.7**, junctions **111.2** (NEW RECORD!)

**Replay Observations:**
- Agents run to edge of map - stuck, cant do anything
- Need reward shaping: penalize being in enemy/neutral territory
- 2 junctions aligned BUT started with 2 hearts in base (not crafted)
- **Next breakthrough:** Can agents craft 3rd heart? Make ONE new heart?

**Code Fixes:**
- Fixed e2e_seconds/api_cost_usd column swap in train.py
- Loop now defers to program.md as single source of truth

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-14 03:25 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-14 05:24 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-14 07:24 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-14 09:24 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-14 17:42 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-14 20:16 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-14 22:12 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.

**Comment by SolbiatiAlessandro** [comment]
🤖 Tashi heartbeat check (2026-03-14 23:43 PT): Noticed no new commits in a while. Is the session still running? If stuck, check run.log for errors.
